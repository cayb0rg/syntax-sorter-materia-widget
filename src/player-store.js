import React, { useReducer } from 'react'

const init = {
	title: 'New Foreign Language Widget',
	items: [],
	legend: [],
	currentIndex: 0,
	requireInit: true,
	showTutorial: true,
	showWarning: false,
	numAsk: 1,
	askLimit: "no",
	questionsAsked: []
}

const store = React.createContext(init)
const { Provider } = store
let numQuestions = 0

const shuffle = (array) => {
	for (let i = array.length - 1; i > 0; i--) {
		const j = Math.floor(Math.random() * (i + 1));
		[array[i], array[j]] = [array[j], array[i]];
	}
	return array
}

const randTokenOrder = (reals, fakes) => {
	let tLen = reals.length + fakes.length
	let tArr = reals.concat(fakes)
	let indxArr = []
	let orderArr = []

	for (let i = 0; i < tLen; i++)
	{
		indxArr.push(i)
	}

	for (let i = 0; i < tLen; i++)
	{
		let rand = Math.floor(Math.random() * indxArr.length);

		if (indxArr[rand] != undefined)
		{
			orderArr.push(indxArr[rand])
			indxArr.splice(rand, 1)
		}
	}

	return orderArr
}

const importFromQset = (qset) => {
	return {
		items: qset.items.map((item) => {
			let fakes = item.options.fakeout ? shuffle(item.options.fakeout.map((token) => { return {...token, status: 'unsorted', fakeout: true}})) : []
			let reals = shuffle(item.answers[0].options.phrase.map((token) => { return {...token, status: 'unsorted', fakeout: false} }))
			return {
				question: item.questions[0].text,
				answer: item.answers[0].text,
				hint: item.options.hint,
				fakeout: fakes,
				phrase: reals,
				sorted: [],
				displayPref: item.options.displayPref,
				checkPref: item.options.checkPref,
				numChecks: item.options.numChecks,
				fakeoutPref: item.options.fakeoutPref,
				checksUsed: 0,
				correct: "none",
				correctPhrase: item.answers[0].options.phrase,
				allFakes: fakes,
				allReals: reals,
				tokenOrder: [],
				qsetId: item.id
			}
		}),
		legend: qset.options.legend,
		numAsk: qset.options.numAsk,
		askLimit: qset.options.askLimit,
		questionsAsked: []
	}
}

const reduceNumAsk = (imported) => {

	let items = imported.items

	if (imported.askLimit == "yes")
	{
		let cloneImported = JSON.parse(JSON.stringify(imported)) // Makes a deep copy
		let newItems = []
		let indxArr = []

		// Initalizes the index array
		for (let i = 0; i < items.length; i++)
		{
			indxArr.push(i)
		}

		// Gets the random questions to ask
		for (let i = 0; i < imported.numAsk; i++)
		{
			let rand = Math.floor(Math.random() * indxArr.length); // [0 - length-1]

			if (items[rand])
			{
				cloneImported.questionsAsked.push(indxArr[rand])
				newItems.push(items[indxArr[rand]])
				indxArr.splice(rand,1) // Gets rid of 1 number at the rand index
			}
			else
			{
				i--
				continue
			}
		}

		cloneImported.items = newItems

		return cloneImported
	}
	else
	{
		return imported
	}
}

const tokenSortedPhraseReducer = (list, action) => {
	switch (action.type) {
		case 'token_dragging':
			return list.map((token, index) => {
				if (action.payload.tokenIndex == index) {
					return {
						...token,
						status: 'dragging'
					}
				}
				else return token
			})
		case 'token_drag_complete':
			return list.map((token, index) => {
				if (action.payload.tokenIndex == index) {
					return {
						...token,
						status: action.payload.origin
					}
				}
				else return token
			})
		case 'sorted_right_click':
			return [
				...list.slice(0, action.payload.tokenIndex),
				...list.slice(action.payload.tokenIndex + 1)
			]
		case 'token_update_position':
			return list.map((token, index) => {
				if (action.payload.tokenIndex == index) {
					return {
						...token,
						position: {
							x: action.payload.x,
							width: action.payload.width
						}
					}
				}
				else return token
			})
		case 'response_token_sort':
			return [
					...list.slice(0, action.payload.targetIndex),
					{
						legend: action.payload.legend,
						value: action.payload.value,
						status: 'sorted',
						fakeout: action.payload.fakeout,
						position: {},
						arrangement: null
					},
					...list.slice(action.payload.targetIndex)
				]
		case 'response_token_rearrange':
			let target = action.payload.targetIndex
			if (action.payload.originIndex < target) target--

			let stageOne = [
				...list.slice(0, action.payload.originIndex),
				...list.slice(action.payload.originIndex + 1)
			]

			return [
				...stageOne.slice(0, target),
				{
					legend: action.payload.legend,
					value: action.payload.value,
					status: 'sorted',
					fakeout: action.payload.fakeout,
					position: {},
					arrangement: null
				},
				...stageOne.slice(target)
			]
		case 'adjacent_token_update':
			return list.map((token) => {

				if (action.payload.left != undefined && token.index == action.payload.left) {
					return {
						...token,
						arrangement: 'left'
					}
				}
				else if (action.payload.right != undefined && token.index == action.payload.right) {
					return {
						...token,
						arrangement: 'right'
					}
				}
				else return {
					...token,
					arrangement: null
				}
			})
		default:
			throw new Error(`Sorted Token phrase reducer: action type: ${action.type} not found.`)
	}
}

const tokenUnsortedPhraseReducer = (list, action) => {
	switch (action.type) {
		case 'token_dragging':
			return list.map((token, index) => {
				if (action.payload.tokenIndex == index) {
					return {
						...token,
						status: 'dragging'
					}
				}
				else return token
			})
		case 'token_drag_complete':
			return list.map((token, index) => {
				if (action.payload.tokenIndex == index) {
					return {
						...token,
						status: action.payload.origin
					}
				}
				else return token
			})
		case 'response_token_sort':
			return [
				...list.slice(0, action.payload.phraseIndex),
				...list.slice(action.payload.phraseIndex + 1)
			]
		case 'sorted_right_click':
			return [
				...list, {
					legend: action.payload.legend,
					value: action.payload.value,
					status: 'unsorted',
					fakeout: action.payload.fakeout
				}
			]
		default:
			throw new Error(`Token phrase reducer: action type: ${action.type} not found.`)
	}
}

const questionItemReducer = (items, action) => {
	switch (action.type) {
		case 'token_dragging':
			return items.map((item, index) => {
				if (index == action.payload.questionIndex) {
					if (action.payload.status == 'unsorted')
					{
						if (action.payload.fakeout == true)
						{
							return { ...item, fakeout: tokenUnsortedPhraseReducer(item.fakeout, action) }
						}
						else
						{
							return { ...item, phrase: tokenUnsortedPhraseReducer(item.phrase, action) }
						}
					}
					else if (action.payload.status == 'sorted') return {...item, sorted: tokenSortedPhraseReducer(item.sorted, action)}
				}
				else return item
			})
		case 'token_drag_complete':
			return items.map((item, index) => {
				if (index == action.payload.questionIndex) {
					if (action.payload.origin == 'unsorted')
					{
						if (action.payload.fakeout == true)
						{
							return { ...item, fakeout: tokenUnsortedPhraseReducer(item.fakeout, action) }
						}
						else
						{
							return { ...item, phrase: tokenUnsortedPhraseReducer(item.phrase, action) }
						}
					}
					else if (action.payload.origin == 'sorted') return {...item, sorted: tokenSortedPhraseReducer(item.sorted, action)}
				}
				else return item
			})
		case 'sorted_right_click':
			return items.map((item, index) => {
				if (index == action.payload.questionIndex) {

					if (action.payload.fakeout == true)
					{
						return { ...item, sorted: tokenSortedPhraseReducer(item.sorted, action), fakeout: tokenUnsortedPhraseReducer(item.fakeout, action) }
					}
					else
					{
						return { ...item, sorted: tokenSortedPhraseReducer(item.sorted, action), phrase: tokenUnsortedPhraseReducer(item.phrase, action) }
					}

					//else if (action.payload.origin == 'sorted') return {...item, sorted: tokenSortedPhraseReducer(item.sorted, action)}
				}
				else return item
			})
		case 'response_token_sort':
			return items.map((item, index) => {
				if (index == action.payload.questionIndex) {
					if (action.payload.fakeout == true)
					{
						return {...item, sorted: tokenSortedPhraseReducer(item.sorted, action), fakeout: tokenUnsortedPhraseReducer(item.fakeout, action)}
					}
					else
					{
						return {...item, sorted: tokenSortedPhraseReducer(item.sorted, action), phrase: tokenUnsortedPhraseReducer(item.phrase, action)}
					}
				}
				else return item
			})
		case 'response_token_rearrange':
		case 'token_update_position':
		case 'adjacent_token_update':
			return items.map((item, index) => {
				if (index == action.payload.questionIndex) {
					return {...item, sorted: tokenSortedPhraseReducer(item.sorted, action)}
				}
				else return item
			})
		case 'correct_update':
			return items.map((item, index) => {
				if (index == action.payload.questionIndex) {
					return {...item, correct: action.payload.answer, checksUsed: (item.checksUsed + 1)}
				}
				else return item
			})
		default:
			throw new Error(`Question item reducer: action type: ${action.type} not found.`)
	}
}

const StateProvider = ( { children } ) => {
	const [state, dispatch] = useReducer((state, action) => {
		switch (action.type) {
			case 'init':
				let imported = importFromQset(action.payload.qset)
				let importedReduced = reduceNumAsk(imported)
				for (let i = 0; i < importedReduced.items.length; i++)
				{
					importedReduced.items[i].tokenOrder = randTokenOrder(importedReduced.items[i].phrase, importedReduced.items[i].fakeout)
				}
				return {...state, title: action.payload.title, items: importedReduced.items, legend: imported.legend, questionsAsked: importedReduced.questionsAsked, requireInit: false}
			case 'toggle_tutorial':
				return {...state, showTutorial: !state.showTutorial}
			case 'toggle_warning':
				return {...state, showWarning: !state.showWarning}
			case 'select_question':
				return {...state, currentIndex: action.payload}
			case 'paginate_question_forward':
				//console.log('current indx: ' + state.currentIndex)
				let forward = state.currentIndex < state.items.length -1 ? state.currentIndex + 1: state.currentIndex
				return {...state, currentIndex: forward}
			case 'token_dragging':
			case 'token_drag_complete':
			case 'sorted_right_click':
			case 'token_update_position':
			case 'response_token_sort':
			case 'response_token_rearrange':
			case 'adjacent_token_update':
			case 'correct_update':
				return {...state, items: questionItemReducer(state.items, action)}
			default:
				throw new Error(`Base reducer: action type: ${action.type} not found.`)
		}
	}, init)

	return <Provider value={{state, dispatch}}>{children}</Provider>
}

export {store, StateProvider }
