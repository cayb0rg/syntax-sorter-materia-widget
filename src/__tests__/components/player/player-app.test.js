import React from 'react'
import renderer from 'react-test-renderer'
import PlayerApp from '../../../components/player/player-app'
import  { StateProvider } from '../../../player-store'

describe('Player Application', () => {

	const title = 'Player App Test Title'
	const _qset = {
		items: [
			{
				questions: [{ text: 'Test Question Text 1'}],
				answers: [{ 
					options: {
						phrase: [
							{
								value: 'This',
								legend: 1
							},
							{
								value: 'is',
								legend: 2
							},
							{
								value: 'a',
								legend: 1
							},
							{
								value: 'test',
								legend: 2
							}
						]
					}
				}],
				options: {
					displayPref: 'word',
					attempts: '1',
					hint: 'Test Question Hint',
					fakes: []
				}
			}
		],
		options: {
			legend: [
				{
					id: 1,
					color: '#FF0000',
					name: 'Test noun',
					focus: false
				},
				{
					id: 2,
					color: '#00FF00',
					name: 'Test verb',
					focus: false
				}
			],
			enableQuestionBank: false,
			requireAllQuestions: true,
			numAsk: 1
		}
	}

	test('Player App component', () => {
		const component = renderer.create(
			<StateProvider>
				<PlayerApp title={title} qset={_qset} />
			</StateProvider>
		)

		const tree = component.toJSON()
		expect(tree).toMatchSnapshot()
	})
})