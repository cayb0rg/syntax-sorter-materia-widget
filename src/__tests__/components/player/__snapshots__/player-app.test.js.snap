// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Player Application Player App component 1`] = `
<div
  className="player-container"
  onKeyDown={[Function]}
>
  <div
    aria-live="assertive"
    id="aria-live-region"
    role="region"
  >
    <p>
      
    </p>
  </div>
  <div
    className="warning-wrapper"
    style={
      Object {
        "display": "none",
      }
    }
  >
    <div
      aria-describedby="warning-desc"
      aria-labelledby="alert-label"
      aria-modal="true"
      className="warning"
      role="alertdialog"
    >
      <div
        id="alert-label"
      >
        <p
          id="dev-warning"
        >
           You still have unfinished questions.
        </p>
        <h3>
          Remaining Questions: 
        </h3>
        <div
          id="remaining-questions"
        />
      </div>
      <div>
        <h3>
          Are you sure you want to submit?
        </h3>
        <div
          className="warning-submit-holder"
        >
          <button
            id="warning-submit-btn"
            onClick={[Function]}
          >
            No
          </button>
          <button
            onClick={[Function]}
          >
            Yes
          </button>
        </div>
      </div>
    </div>
    <div
      className="warning-bg"
    />
  </div>
  <div
    className="modal-wrapper"
    style={
      Object {
        "display": "flex",
      }
    }
  >
    <div
      aria-describedby="dialog-desc"
      aria-labelledby="dialog-title"
      aria-modal="true"
      className="modal tutorial player"
      role="alertdialog"
    >
      <h3
        id="dialog-title"
      >
        Arrange Tokens in the Proper Order
      </h3>
      <div
        id="dialog-desc"
      >
        <div
          className="tutorial-content"
        >
          <p>
            Read each question prompt and arrange the tokens based on the instructions provided. Tokens may be a word, part of a word, or a part of speech. 
            <span
              className="strong"
            >
              Right-click
            </span>
             or 
            <span
              className="strong"
            >
              drag
            </span>
             a token back to the drawer to unsort it. 
          </p>
          <img
            alt="Player tutorial image demonstrating dragging and dropping tokens"
            className="right"
            src="assets/img/player-tutorial-1.png"
          />
        </div>
        <div
          className="tutorial-content"
        >
          <p>
            If using a keyboard, select a token in the drawer and press 
            <span
              className="strong"
            >
              Space
            </span>
             or 
            <span
              className="strong"
            >
              Enter
            </span>
             to sort it. To rearrange a token in the sorting area, press 
            <span
              className="strong"
            >
              Q
            </span>
             to move it left and 
            <span
              className="strong"
            >
              E
            </span>
             to move it right.
          </p>
        </div>
        <div
          className="tutorial-content"
        >
          <p>
            To read out the tokens you have sorted, press 
            <span
              className="strong"
            >
              R
            </span>
             or click the 
            <span
              className="strong"
            >
              Speaker
            </span>
             on the bottom right of the sorting area.
          </p>
        </div>
        <div
          className="tutorial-content"
        >
          <img
            alt="Player tutorial image demonstrating using the color legend."
            className="left"
            src="assets/img/player-tutorial-2.png"
          />
          <p>
            Tokens are color-coded based on their syntax. Use the Legend at the bottom to associate each color with its corresponding syntax.
          </p>
        </div>
        <div
          className="tutorial-content"
        >
          <p>
            Some questions may allow you to check your answer multiple times before submitting for a score. Use the 
            <span
              className="strong"
            >
              Check Answer
            </span>
             button to see how you did!
          </p>
          <img
            alt="Player tutorial image demonstrating using the Check Answer feature."
            className="right"
            src="assets/img/player-tutorial-3.png"
          />
        </div>
        <div
          className="tutorial-content"
        >
          <p>
            Some questions may have hints. Press 
            <span
              className="strong"
            >
              H
            </span>
             to play the hint if it's available.
          </p>
        </div>
      </div>
      <button
        id="tutorial-ready-btn"
        onClick={[Function]}
      >
        I'm Ready
      </button>
    </div>
    <div
      className="modal-bg"
    />
  </div>
  <header
    aria-hidden="true"
    className="player-header"
    inert=""
  >
    <h1
      className="title"
    >
      New Syntax Sorter Widget
    </h1>
    <div
      className="player-header-btns"
    >
      <button
        className="headerBtn"
        onClick={[Function]}
      >
        Tutorial
      </button>
      <button
        className="headerBtn"
        onClick={[Function]}
      >
        Submit
      </button>
    </div>
  </header>
  <nav
    aria-hidden="true"
    className="question-select"
    inert=""
  >
    <button
      className="select-btn paginate-up  disabled"
      disabled={true}
      onClick={[Function]}
    >
      <span
        className="icon-arrow-up2"
      />
    </button>
    <button
      className="select-btn paginate-down  disabled"
      disabled={true}
      onClick={[Function]}
    >
      <span
        className="icon-arrow-down2"
      />
    </button>
  </nav>
  <main
    aria-hidden="true"
    className="content-container"
    inert=""
  >
    <section
      className="card question-container"
    >
      <h2
        aria-label="Question: Drag and drop to arrange the items below in the correct order."
        id="question-text"
      >
        Drag and drop to arrange the items below in the correct order.
      </h2>
      <div
        aria-live="assertive"
        className="hint-text "
      >
        <p>
          <span
            className="strong"
          >
            Hint: 
          </span>
          <span />
        </p>
      </div>
    </section>
    <section
      className="card phrase-player undefined "
    >
      <div
        className="token-container "
      >
        <div
          className="token-target"
          onDragOver={[Function]}
          onDrop={[Function]}
        >
          Drag and drop the words below to arrange them. If using a keyboard, select a token in the drawer and press Space or Enter to sort it. To rearrange a token in the sorting area, press Q to move it left and E to move it right.
        </div>
        <button
          aria-label="Read Sorted Phrase"
          id="play-audio-btn"
          onClick={[Function]}
          onMouseOut={[Function]}
          onMouseOver={[Function]}
          title="Read Sorted Phrase"
        >
          <img
            aria-hidden="true"
            src="./assets/img/volume-medium.svg"
          />
        </button>
        <span
          className="fakeout-tip "
        >
          <span
            className="icon-notification"
          />
          Not all of the items below may be part of the correct phrase.
        </span>
      </div>
      <section
        className="token-drawer undefined "
        onDragOver={[Function]}
        onDrop={[Function]}
      >
        <div
          aria-describedby="token-drawer-desc"
          role="group"
        >
          <h3
            id="token-drawer-desc"
          >
            Token Drawer
          </h3>
        </div>
        <section
          className="response-controls"
        >
          <div
            className="response-message-container"
          >
            <div
              id="response-dialog-desc"
            >
              <span>
                NONE
              </span>
            </div>
          </div>
          <div
            className="button-container"
          >
            <button
              aria-describedby="response-dialog-desc"
              className="verify "
              id="check-question-btn"
              onClick={[Function]}
            >
              Check Answer
            </button>
            <button
              aria-describedby="response-dialog-desc"
              className="paginate show"
              id="next-question-btn"
              onClick={[Function]}
            >
              Next Question
            </button>
          </div>
        </section>
      </section>
    </section>
    <section
      className="card legend"
    >
      <header
        id="color-legend-header"
      >
        Color Legend
      </header>
      <dl
        aria-labelledby="color-legend-header"
      />
    </section>
  </main>
</div>
`;
